pkg/cri/store/stats/stats.go

```diff
@@ -1,27 +1,71 @@
-/*
-   Copyright The containerd Authors.
+package stats
 
-   Licensed under the Apache License, Version 2.0 (the "License");
-   you may not use this file except in compliance with the License.
-   You may obtain a copy of the License at
+import (
+	runtime "k8s.io/cri-api/pkg/apis/runtime/v1"
+)
 
-       http://www.apache.org/licenses/LICENSE-2.0
+// ContainerStats contains the information about container stats.
+type ContainerStats struct {
+	Attributes      *ContainerAttributes
+	CPUStats        *ContainerCpuStats
+	MemoryStats     *ContainerMemoryStats
+	FileSystemStats *ContainerFileSystemStats
+	CPUStatsUpdate  *ContainerCpuStatsUpdate
+}
 
-   Unless required by applicable law or agreed to in writing, software
-   distributed under the License is distributed on an "AS IS" BASIS,
-   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
-   See the License for the specific language governing permissions and
-   limitations under the License.
-*/
+type ContainerCpuStats struct {
+	Timestamp int64
+	// Cumulative CPU usage (sum across all cores) since object creation.
+	UsageCoreNanoSeconds uint64
+	// Total CPU usage (sum of all cores) averaged over the sample window.
+	// The "core" unit can be interpreted as CPU core-nanoseconds per second.
+	UsageNanoCores uint64
+}
 
-package stats
+type ContainerMemoryStats struct {
+	Timestamp int64
+	// The amount of working set memory in bytes.
+	WorkingSetBytes uint64
+	// Available memory for use. This is defined as the memory limit - workingSetBytes.
+	AvailableBytes uint64
+	// Total memory in use. This includes all memory regardless of when it was accessed.
+	UsageBytes uint64
+	// The amount of anonymous and swap cache memory (includes transparent hugepages).
+	RssBytes uint64
+	// Cumulative number of minor page faults.
+	PageFaults uint64
+	// Cumulative number of major page faults.
+	MajorPageFaults uint64
+}
 
-import "time"
+type ContainerFileSystemStats struct {
+	Timestamp int64
+	// The unique identifier of the filesystem.
+	FsId FilesystemIdentifier
+	// UsedBytes represents the bytes used for images on the filesystem.
+	// This may differ from the total bytes used on the filesystem and may not
+	// equal CapacityBytes - AvailableBytes.
+	UsedBytes uint64
+	// InodesUsed represents the inodes used by the images.
+	// This may not equal InodesCapacity - InodesAvailable because the underlying
+	// filesystem may also be used for purposes other than storing images.
+	InodesUsed uint64
+}
 
-// ContainerStats contains the information about container stats.
-type ContainerStats struct {
-	// Timestamp of when stats were collected
-	Timestamp time.Time
+type FilesystemIdentifier struct {
+	// Mountpoint of a filesystem.
+	Mountpoint string
+}
+
+type ContainerAttributes struct {
```

Comment:
hmm, why do we need this? Isn't already stored in the containerd store? I was under the assumption we already maintain a map of container id -> container stats so we don't need to store this info (it can be looked up on the fly if needed).

Reply:
I believe there are some discrepancies between the cri and containerd store info. E.g. lack of `Attempt` which is present in `ContainerMetadata` and potentially lack of annotations

Reply:
Hmm, isn't all this info stored in the container store? If it's the case, why do we need to store it here again. For example, for annotations, they can be pulled from the metadata object on the container like so: https://github.com/containerd/containerd/blob/main/pkg/cri/sbserver/container_stats_list_windows.go#L56

Regarding attempt, where does that info come from, can it be looked up as well?

Reply:
Updated